utils/rjson.cc

```diff
@@ -236,6 +240,67 @@ std::string print(const rjson::value& value, size_t max_nested_level) {
     return std::string(buffer.GetString());
 }
 
+future<> print(const rjson::value& value, seastar::output_stream<char>& os, size_t max_nested_level) {
+    struct os_buffer {
+        seastar::output_stream<char>& _os;
+        temporary_buffer<char> _buf;
+        size_t _pos = 0;
+        future<> _f = make_ready_future<>();
+
+        using Ch = char;
+
+        void send(bool try_reuse = true) {
+            if (_f.failed()) {
+                _f.get0();
+            }
+            if (!_buf.empty() && _pos > 0) {
+                _buf.trim(_pos);
+                _pos = 0;
+                // Note: we're assuming we're writing to a buffered output_stream (hello http server).
+                // If we were not, or if (http) output_stream supported mixed buffered/packed content
+                // it might be a good idea to instead send our buffer as a packet directly. If so, the
+                // buffer size should probably increase (at least after first send()).
+                _f = _f.then([this, buf = std::move(_buf), &os = _os, try_reuse]() mutable -> future<> {
+                    return os.write(buf.get(), buf.size()).then([this, buf = std::move(buf), try_reuse]() mutable {
+                        // Chances are high we just copied this to output_stream buffer, and got here
+                        // immediately. If so, reuse the buffer.
+                        if (try_reuse && _buf.empty() && _pos == 0) {
+                            _buf = std::move(buf);
+                        }
+                    });
+                });
+            }
+        }
+        void Put(char c) {
+            if (_pos == _buf.size()) {
+                send();
+                if (_buf.empty()) {
+                    _buf = temporary_buffer<char>(512);
+                }
+            }
+            // Second note: Should consider writing directly to the buffer in output_stream
+            // instead of double buffering. But output_stream for a single char has higher
+            // overhead than the above check + once we hit a non-completed future, we'd have
+            // to revert to this method anyway...
+            *(_buf.get_write() + _pos) = c;
+            ++_pos;
+        }
+        void Flush() {
+            send();
+        }
+        future<> finish()&& {
+            send(false);
+            return std::move(_f);
+        }
+    };
+
+    os_buffer osb{ os };
+    using streamer = rapidjson::Writer<os_buffer, encoding, encoding, allocator>;
+    guarded_yieldable_json_handler<streamer, false, os_buffer> writer(osb, max_nested_level);
+    value.Accept(writer);
+    co_return co_await std::move(osb).finish();
```

Comment:
I think this std::move() and "&&" on the function is not necessary (and might confuse a reader to think it does something), but it doesn't hurt anything either, so ok.

Reply:
It is not strictly required, but since this is a fundamentally state-altering function I think it makes things clearer. `finish()` is for all intents and purposes our future<>-returning destructor.